### PR TITLE
Add DAG run abort endpoint

### LINKED ISSUE #15888: Abort a DAG Run
**Description**

It would be great having a option to abort a DAG Runs through the REST API.

**Use case / motivation**

The proposed input params would be:
 - DAG_ID
 - DAG_RUN_ID

The DAG Run should abort all its tasks running and mark them as "failed".

**Are you willing to submit a PR?**

**Related Issues**

### PR DESCRIPTION
**Pull request by bbenshalom**
Fixes #15888 

The new endpoint will set the DAG run state and all of the task instances' state to FAILED
I couldn't find any tests for the endpoints directly, so did not add a unit test... Please let me know if I missed their location and I'll happily add them!

### CODE REVIEW

**Review by ephraimbuddy — CHANGES_REQUESTED** [review]
What we should do here is give the ability to mark DagRun success or failed just as we do in the UI, in one endpoint. 
Also, tests for dag_run_endpoint can be found here: https://github.com/apache/airflow/blob/main/tests/api_connexion/endpoints/test_dag_run_endpoint.py

File: airflow/api_connexion/endpoints/dag_run_endpoint.py
Comment (uranusjr) [inline airflow/api_connexion/endpoints/dag_run_endpoint.py]:
```suggestion
    dag_run: Optional[DagRun] = session.query(DagRun).filter(DagRun.dag_id == dag_id, DagRun.run_id == dag_run_id).one_or_none()
```

File: airflow/api_connexion/endpoints/dag_run_endpoint.py
Comment (uranusjr) [inline airflow/api_connexion/endpoints/dag_run_endpoint.py]: This should happen before loading `post_body`. If a DAG does not exist, the endpoint should always fail with 404.

File: airflow/api_connexion/openapi/v1.yaml
Comment (uranusjr) [inline airflow/api_connexion/openapi/v1.yaml]: This endpoint feels wrong to me. Since the operation only affects one single DagRun, I feel it should be something like `/dags/{dag_id}/dagRuns/{dag_run_id}/state`. And it should only take a single `state` field (not `dag_id` or `dag_run_id`).

**Review by uranusjr — CHANGES_REQUESTED** [review]

**Comment by uranusjr** [comment]
Also I think the linter will reject your code style. Please considering setting up pre-commit and check for the formatting issues locally; see `CONTRIBUTING.rst` for details.

File: airflow/api_connexion/endpoints/dag_run_endpoint.py
Comment (uranusjr) [inline airflow/api_connexion/endpoints/dag_run_endpoint.py]:
```suggestion
    dag_run: Optional[DagRun] = session.query(DagRun).filter(DagRun.dag_id == dag_id, DagRun.run_id == dag_run_id).one_or_none()
    if dag_run is None:
        error_message = f'Dag Run id {dag_run_id} not found in dag {dag_id}'
        raise DagRunNotFound(error_message)
    try:
        post_body = dagrun_schema.load(request.json, session=session)
    except ValidationError as err:
        raise BadRequest(detail=str(err))
```

Also, is the `unknown="include"` argument needed and why?

**Review by uranusjr — CHANGES_REQUESTED** [review]

**Comment by bbenshalom** [comment]
@uranusjr I removed the dagrun_schema.load() part, as I think I don't actually need it, since I already query the dag run and merge it... correct me if I'm wrong

File: airflow/api_connexion/schemas/dag_run_schema.py
Comment (uranusjr) [inline airflow/api_connexion/schemas/dag_run_schema.py]:
```suggestion
class SetDagRunStateFormSchema(Schema):
    """Schema for handling the request of setting state of DAG run"""

    state = DagStateField()
```

We don’t have these fields in the schema now. Also `dump_only` means this field cannot be include in a POST form, which is opposite to what you want. Technically you can add `load_only=True`, but since this entire schema is only used to load data anyway, the flag would not practically change anything.

**Review by uranusjr — CHANGES_REQUESTED** [review]
Also you will still need to satisfy the linter. Again, please consider setting up `pre-commit`.

**Comment by uranusjr** [comment]
> I removed the `dagrun_schema.load()` part, as I think I don't actually need it, since I already query the dag run and merge it... correct me if I'm wrong

The `load` call performs data sanitation and automatically emits a 400 Bad Request if the user POSTs an invalid state name (say `{"state": "asdfg"}`). You need to keep it.

File: airflow/api_connexion/schemas/dag_run_schema.py
Comment (bbenshalom) [reply airflow/api_connexion/schemas/dag_run_schema.py]: > Also you will still need to satisfy the linter. Again, please consider setting up `pre-commit`.

ran the static checks locally, I think it should be good now (for some reason they don't run automatically here in the GitHub actions), thanks!

File: airflow/api_connexion/endpoints/dag_run_endpoint.py
Comment (ephraimbuddy) [inline airflow/api_connexion/endpoints/dag_run_endpoint.py]:
```suggestion
def update_dag_run_state(dag_id: str, dag_run_id: str, session) -> dict:
```

File: airflow/api_connexion/openapi/v1.yaml
Comment (ephraimbuddy) [inline airflow/api_connexion/openapi/v1.yaml]:
```suggestion
          description: The state to set this DagRun
          type: string
          enum:
              - failed
              - success
```
Instead of using the DagState, we should use enums to avoid setting DagRun state to running or queued which are DagStates

File: airflow/api_connexion/openapi/v1.yaml
Comment (ephraimbuddy) [inline airflow/api_connexion/openapi/v1.yaml]:
```suggestion
      operationId: update_dag_run_state
```

File: tests/api_connexion/endpoints/test_dag_run_endpoint.py
Comment (ephraimbuddy) [inline tests/api_connexion/endpoints/test_dag_run_endpoint.py]: We should parametrize this and make sure that DagRuns are not set to `queued` or `running`

**Review by ephraimbuddy — CHANGES_REQUESTED** [review]

File: airflow/api_connexion/endpoints/dag_run_endpoint.py
Comment (uranusjr) [inline airflow/api_connexion/endpoints/dag_run_endpoint.py]: I think this still needs a conversion

```suggestion
    dag_run.set_state(state=DagRunState(state))
```

File: airflow/api_connexion/schemas/dag_run_schema.py
Comment (uranusjr) [inline airflow/api_connexion/schemas/dag_run_schema.py]: I think this should be `Schema`; this is not a SQLAlchemy model.

File: airflow/api_connexion/endpoints/dag_run_endpoint.py
Comment (uranusjr) [inline airflow/api_connexion/endpoints/dag_run_endpoint.py]:
```suggestion
    dag_run.set_state(state=DagRunState(state))
```

`set_state` accepts a `DagRunState`, not `str`.

File: airflow/api_connexion/openapi/v1.yaml
Comment (uranusjr) [inline airflow/api_connexion/openapi/v1.yaml]:
```suggestion
      description: Set a state of DAG run
```

(Nitpicking) don’t need to extra wrap.

File: airflow/api_connexion/openapi/v1.yaml
Comment (uranusjr) [inline airflow/api_connexion/openapi/v1.yaml:2036]: I wonder if it makes sense to set the state to other valid values. But I guess those can be added later if someone asks for them, so they should not block this PR.

File: airflow/api_connexion/schemas/dag_run_schema.py
Comment (uranusjr) [inline airflow/api_connexion/schemas/dag_run_schema.py]: Nitpick: Use `DagRunState.SUCCESS.value` (and `FAILED`) instead.

File: tests/api_connexion/endpoints/test_dag_run_endpoint.py
Comment (uranusjr) [inline tests/api_connexion/endpoints/test_dag_run_endpoint.py]: It’s probably easier to use the `dag_maker` fixture for this (but the current implementation is OK).

**Review by uranusjr — CHANGES_REQUESTED** [review]

File: airflow/api_connexion/openapi/v1.yaml
Comment (ephraimbuddy) [reply airflow/api_connexion/openapi/v1.yaml:2036]: The other values are `queued` and `running` which are handled by SchedulerJob, so I think it's OK and I agree with you, we should leave it until the need arises

File: tests/api_connexion/endpoints/test_dag_run_endpoint.py
Comment (ephraimbuddy) [inline tests/api_connexion/endpoints/test_dag_run_endpoint.py]:
```suggestion
    def test_should_respond_200(self, state, dag_maker):
        dag_id = "TEST_DAG_ID"
        dag_run_id = 'TEST_DAG_RUN_ID'
        with dag_maker(dag_id):
            DummyOperator(task_id='task_id')
        dag_maker.create_dagrun(run_id=dag_run_id)
```

File: tests/api_connexion/endpoints/test_dag_run_endpoint.py
Comment (ephraimbuddy) [inline tests/api_connexion/endpoints/test_dag_run_endpoint.py]:
```suggestion
            f"api/v1/dags/{dag_id}/dagRuns/{dag_run_id}/state",
```

File: tests/api_connexion/endpoints/test_dag_run_endpoint.py
Comment (ephraimbuddy) [inline tests/api_connexion/endpoints/test_dag_run_endpoint.py]:
```suggestion
            'dag_id': dag_id,
            'dag_run_id': dag_run_id,
```

**Review by ephraimbuddy — CHANGES_REQUESTED** [review]
Cool thing using dag_maker in your tests. Take a look at my suggestions and apply same to the other test

File: tests/api_connexion/endpoints/test_dag_run_endpoint.py
Comment (bbenshalom) [reply tests/api_connexion/endpoints/test_dag_run_endpoint.py]: @ephraimbuddy I can't get this setup to work and don't really understand it, I guess, so a couple of questions:
1. How can I inject dag_maker without using `@pytest.fixture`, if I'm also using `@parameterized.expand`? I get an error that I'm missing an error `missing 1 required positional argument: 'dag_maker'`
2. What is the purpose of DummyOperator here, and why run in separately in a context manager from `dag_maker.create_dagrun()`?
I'm considering reverting the usage of this and just creating the DAG using the constructor, like the other tests in this file...

Comment (uranusjr) [reply tests/api_connexion/endpoints/test_dag_run_endpoint.py]: > How can I inject dag_maker […]?

Just include `dag_maker` as an argument. Pytest magically picks the argument up and injects the fixture into the function. But to make the test compatible with pytest, you need to use `pytest.mark.parametrize` instead of `parameterized.expand`.

So something like

```python
@pytest.mark.parametrize("state", [...])
def test_should_respond_200(self, dag_maker, state):
    ...
```

> What is the purpose of DummyOperator here, and why run in separately in a context manager from `dag_maker.create_dagrun()`?

The DummyOperator will be included as a part of the DAG, similar to

```python
with DAG(...):
    MyOperator(...)

# The above is equivalent to
# dag = DAG(...)
# MyOperator(..., dag=dag)
```

It’s not actually used anywhere in the test, so technically you can just do

```python
with dag_maker(...):
    pass
```

A DAG without any operators is kind of weird, but allowed.

Comment (bbenshalom) [reply tests/api_connexion/endpoints/test_dag_run_endpoint.py]: Thanks for the explanation! Makes sense now.

**Review by uranusjr — APPROVED** [review]

File: airflow/api_connexion/openapi/v1.yaml
Comment (mik-laj) [inline airflow/api_connexion/openapi/v1.yaml]:
```suggestion
  /dags/{dag_id}/dagRuns/{dag_run_id}:
    parameters:
      - $ref: '#/components/parameters/DAGID'
      - $ref: '#/components/parameters/DAGRunID'

    patch:
```
This endpoint modifies an existing object, so according to the REST convention, we should not create a new endpoint, but change the method type to PATCH.

Comment (mik-laj) [reply airflow/api_connexion/openapi/v1.yaml]: Similar to:
https://github.com/apache/airflow/blob/fb11ba5c42c2547e6550480eca1c5db14ea76f99/airflow/api_connexion/openapi/v1.yaml#L325-L353
https://github.com/apache/airflow/blob/fb11ba5c42c2547e6550480eca1c5db14ea76f99/airflow/api_connexion/openapi/v1.yaml#L454-L479

File: airflow/api_connexion/openapi/v1.yaml
Comment (mik-laj) [inline airflow/api_connexion/openapi/v1.yaml]: Should we accept DAGRun here similar to `PATCH /dags/{dag_id}`?
https://github.com/apache/airflow/blob/fb11ba5c42c2547e6550480eca1c5db14ea76f99/airflow/api_connexion/openapi/v1.yaml#L466

File: airflow/api_connexion/schemas/dag_run_schema.py
Comment (mik-laj) [inline airflow/api_connexion/schemas/dag_run_schema.py:108]: Can we use DAGRunSchema schema here? I think it is possible, but we need to carefully review which fields are marked as input and which are marked as output.

Comment (uranusjr) [reply airflow/api_connexion/schemas/dag_run_schema.py:108]: It should be possible, but I’d say let’s not bother until someone comes up with a feature request 🙂

Comment (uranusjr) [reply airflow/api_connexion/schemas/dag_run_schema.py:108]: Using `PATCH` is a good idea though. I was leaning toward `POST` with the impression `dagrun.state = new_state` does not work, but I was wrong, it actually implicitly calls `set_state`.

File: airflow/api_connexion/openapi/v1.yaml
Comment (ephraimbuddy) [reply airflow/api_connexion/openapi/v1.yaml]: It will give room to updating the state to any value, and the state on DAGRun is readonly, making it accept a value will lead to setting the DAGrun to running in the API

File: airflow/api_connexion/openapi/v1.yaml
Comment (bbenshalom) [reply airflow/api_connexion/openapi/v1.yaml]: done

File: airflow/api_connexion/openapi/v1.yaml
Comment (bbenshalom) [reply airflow/api_connexion/openapi/v1.yaml]: considering @ephraimbuddy 's comment, I'll resolve this conversation without changing the input schema

File: airflow/api_connexion/schemas/dag_run_schema.py
Comment (bbenshalom) [reply airflow/api_connexion/schemas/dag_run_schema.py:108]: considering @uranusjr 's comment, I'll resolve this conversation without changing the super class here (let me know if you still want it to be done)

**Comment by bbenshalom** [comment]
@ephraimbuddy any additional requested changes I missed? I see that a request from you is blocking the option to merge

File: airflow/api_connexion/openapi/v1.yaml
Comment (uranusjr) [inline airflow/api_connexion/openapi/v1.yaml]:
```suggestion
      summary: Modify a DAG run
      description: Modify a DAG run
```

**Review by uranusjr — CHANGES_REQUESTED** [review]

**Review by uranusjr — APPROVED** [review]

**Review by ephraimbuddy — APPROVED** [review]

**Comment by bbenshalom** [comment]
@uranusjr @ephraimbuddy thank you for the review! Can you please approve the workflow? And since I don't have write access, I guess this will be merged by one of the maintainers when appropriate, right?

**Comment by alex-astronomer** [comment]
Also closes: #17839

**Comment by alex-astronomer** [comment]
Could use more documentation about how to use this endpoint.  Not sure if that exists anywhere, I couldn't find in the `v1.yaml` file.  Had to dig into source code/guess how to use it, maybe I'm just a beginner but I think adding some documentation could go a long way for other beginners.

**Pull request merged**
